core/src/main/java/org/apache/calcite/sql/SqlUtil.java

```diff
@@ -417,24 +418,52 @@ public static void unparseSqlIdentifierSyntax(
     writer.endList(frame);
   }
 
+  private static void unparseSetOpSyntax(
+      SqlNode sqlNode,
+      SqlWriter writer,
+      int leftPrec,
+      int rightPrec) {
+    SqlSelect sqlSelect = (SqlSelect) sqlNode;
+    if (sqlSelect.getOrderList() != null
+        || sqlSelect.getFetch() != null
+        || sqlSelect.getOffset() != null) {
+      final SqlWriter.Frame operandFrame0 = writer.startList(FrameTypeEnum.SETOP, "(", ")");
+      sqlSelect.unparse(writer, leftPrec, rightPrec);
+      writer.endList(operandFrame0);
+    } else {
+      sqlSelect.unparse(writer, leftPrec, rightPrec);
+    }
+  }
+
   public static void unparseBinarySyntax(
       SqlOperator operator,
       SqlCall call,
       SqlWriter writer,
       int leftPrec,
       int rightPrec) {
     assert call.operandCount() == 2;
+    boolean isSetOp = operator instanceof SqlSetOperator;
     final SqlWriter.Frame frame =
         writer.startList(
-            (operator instanceof SqlSetOperator)
+            isSetOp
                 ? SqlWriter.FrameTypeEnum.SETOP
                 : SqlWriter.FrameTypeEnum.SIMPLE);
-    call.operand(0).unparse(writer, leftPrec, operator.getLeftPrec());
+    SqlNode operand0 = call.operand(0);
+    if (isSetOp && operand0.getKind() == SqlKind.SELECT) {
+      unparseSetOpSyntax(operand0, writer, leftPrec, operator.getLeftPrec());
+    } else {
+      operand0.unparse(writer, leftPrec, operator.getLeftPrec());
+    }
     final boolean needsSpace = operator.needsSpace();
     writer.setNeedWhitespace(needsSpace);
     writer.sep(operator.getName());
     writer.setNeedWhitespace(needsSpace);
-    call.operand(1).unparse(writer, operator.getRightPrec(), rightPrec);
+    SqlNode operand1 = call.operand(1);
```

Comment:
Is the logic of the `if` statement below repeated inside the `unparseSetOpSyntax` function?

Reply:
I moved it to `unparseBinaryOperand`.